### PR TITLE
#441: added support for fixing the paramLabel as-is and omit regular processing

### PR DESCRIPTION
**Pull request by bbottema**
Here's a simple fix that allows users to choose to omit regular paramLabel processing and fix it to render as-is instead (ANSI codes notwithstanding).

### CODE REVIEW

File: src/main/java/picocli/CommandLine.java [hunk 1]
```diff
@@ -4332,6 +4340,9 @@ protected int hashCodeImpl() {
                 /** Sets the name of the option or positional parameter used in the usage help message, and returns this builder. */
                 public T paramLabel(String paramLabel)       { this.paramLabel = Assert.notNull(paramLabel, "paramLabel"); return self(); }
     
+                /** Sets the name of the option or positional parameter used in the usage help message, and returns this builder. */
```
Comment (bbottema) [inline src/main/java/picocli/CommandLine.java]: Just noticed I didn't update this javadoc entry...

**Comment by bbottema** [comment]
Note: I went with a boolean flag that changes how `paramLabel` works, rather than adding a `customParamLabel`, otherwise you would have to add more logic to manage priority between the two.

I feel two String properties that affect the same output if is more difficult to explain to the library users than explicitly modifying a single existing String property with a boolean flag.

**Comment by remkop** [comment]
One idea is to call the flag ‘paramLabelDecorations‘ or ‘paramLabelIndindicators‘ (default ‘true‘).

**Comment by bbottema** [comment]
I'm fine with either, although personally I do prefer to use a _verb_ that describes what it does. So something like like `renderParamLabelDecorations `or `renderParamLabelIndindicators` (or `applyXYZ`)

**Comment by remkop** [comment]
Yes, that’s good style. I just like starting with `paramLabel` so they show up together in the IDE. But maybe overthinking it.

**Comment by bbottema** [comment]
Good point. Perhaps something like paramLabelShouldHaveDecorations, but maybe now I am over thinking it :smiley:

File: src/main/java/picocli/CommandLine.java (hunk 1 above)
Comment (remkop) [reply src/main/java/picocli/CommandLine.java]: If you push additional commits to your branch, they will show up in the PR...

File: src/test/java/picocli/CommandLineModelTest.java [hunk 2]
```diff
@@ -66,11 +66,14 @@ public void testModelUsageHelp() throws Exception {
...
         spec.addOption(OptionSpec.builder("-V", "--version").versionHelp(true).description("show help and exit").build());
         spec.addOption(OptionSpec.builder("-c", "--count").paramLabel("COUNT").arity("1").type(int.class).description("number of times to execute").build());
+        spec.addOption(OptionSpec.builder("-f", "--fix").paramLabel("FIXED(=BOOLEAN)").arity("1").fixParamLabel(true).required(true).description("run with fixed option").build());
         CommandLine commandLine = new CommandLine(spec);
         String actual = usageString(commandLine, Ansi.OFF);
         String expected = String.format("" +
-                "Usage: <main class> [-hV] [-c=COUNT]%n" +
+                "Usage: <main class> [-hV] [-c=COUNT] -f FIXED(=BOOLEAN)%n" +
                 "  -c, --count=COUNT   number of times to execute%n" +
+                "  -f, --fix FIXED(=BOOLEAN)%n" +
```
[the remark this answers is not in the document]
Comment (remkop) [reply src/test/java/picocli/CommandLineModelTest.java]: A hardcoded space is incorrect. Picocli currently inserts the value of `commandSpec.parser().separator()`. This behaviour needs to be preserved.

**Comment by remkop** [comment]
As for the name of this switch, I'm thinking `showParamSyntax` may be a good name.

**Comment by bbottema** [comment]
Applied the changes as discussed. The master build is failing though, so this fails as well since I updated the branch to master...

**Comment by remkop** [comment]
Thanks. Changes look good. We should also add this attribute to the `@Option` and `@Parameters` annotations.

**Comment by remkop** [comment]
What do you think of the idea to flip this around and call it `hideParamSyntax` (default `false`) instead? That makes it slightly more clear that this is shown by default.

**Comment by remkop** [comment]
@bbottema  I think we are almost there. There is one test that fails now because the default changed from true to false. Will you be able to add the `hideParamSyntax` attribute to the `@Option` and `@Parameters` annotations?

**Comment by bbottema** [comment]
Took care of the junit test, but I don't have time left to work on it for the moment. Got caught up in a nasty double byte character set [bug](https://github.com/bbottema/outlook-message-parser/issues/3) in a RTF to HTML converter 😮

**Comment by bbottema** [comment]
I did what I could aside from the annotations stuff.

I'm calling it a day on this one because frankly speaking I find it difficult to learn the layers and relationships between the classes having to scroll all the time between them over almost 10k LOC (no overview). I tried a couple of times, but kept putting it down in a mild exasperation. File search is not helping much either since a word like "paramLabel" occurs 82 times when limited to whole words only in just the CommandLine.java, not counting comments. It doesn't help that even IntelliJ has buffering issues with this much code (and I have a monstrous custom built laptop).

Anyway, sorry for the rant. I really like what picocli is capable off, but it is too tedious for me personally to learn the intricacies of the library for a newcomer. I have half a mind of forking it and restructuring things a bit for my own use...

**Comment by remkop** [comment]
Sorry for the late response: traveling with family, intermittent access to internet and PC. 

Thanks for the contributions so far! Picocli is a better library because of your work. 

I realize that the single source file is becoming a stumbling block for contributors. I’ll investigate whether it’s possible to do something in the build: have a Gradle task that picks the source apart into smaller chunks, let people work with these (ideally run the tests against them) and eventually recombine into a single source file again.

**Comment by remkop** [comment]
I need a bit more time before I can get back to this PR.

Meanwhile, I added some class diagrams (https://github.com/remkop/picocli/tree/master/src/main/java/picocli/doc-files). Hope this is useful.

**Pull request merged**